src/librustdoc/html/render/mod.rs

```diff
@@ -3560,6 +3557,12 @@ fn render_assoc_items(
             render_deref_methods(w, cx, impl_, containing_item, has_deref_mut, cache);
         }
 
+        // If we were already one level into rendering deref methods, we don't want to render
+        // anything after recusring into any further deref methods above.
+        if let AssocItemRender::DerefFor { .. } = what {
+            return;
+        }
```

Comment:
The impls were already documented by `render_deref_methods` and we don't want to duplicate them, is that the idea? The condition for running `render_deref` is different from what you check here though: this checks `DerefFor` and render_derf checks `deref_impl.is_some()`. Is there any time those could be different? If not, could you add an assert?

Reply:
There are two concepts at play here simultaneously which is a bit confusing, so I'll try to explain and then we can work out whatever comments or assertions might be helpful.

Let's consider a two level `Deref` case:

```rust
pub struct Foo(Bar);
pub struct Bar(Baz);
pub struct Baz;

impl Deref for Foo {
    type Target = Bar;
    fn deref(&self) -> &Bar { &self.0 }
}

impl Deref for Bar {
    type Target = Baz;
    fn deref(&self) -> &Baz { &self.0 }
}
```

The goal of `if let Some(impl_) = deref_impl` when calling `render_deref_methods` is to **descend** to the methods of the `Deref` target: e.g. from `Foo` to `Bar` and from `Bar` to `Baz`.

The goal of `if let AssocItemRender::DerefFor { .. } = what` is to **restrict** what we render when we have _already_ descended at least one level: e.g. when we're at `Bar` and `Baz`, we only want to show their methods and nothing else.

Does that help at all? 😅

Reply:
Yes, that helps me understand why these are different, thanks :) I still have my original question, although maybe I didn't word it very well. This `return` ignores synthetic impls, which is correct (the `Target` could implement Sized but not the original, and vice-versa), but it also ignores blanket and concrete impls. Where are those handled? Is it in `render_deref_items`?

Reply:
At the moment, this change is preserving the previous behaviour: trait, synthetic, and blanket impls are all skipped for `Deref` targets (e.g. `Bar` and `Baz` in the above example). The methods in the `Deref` target that we want to include are rendered by [the top bit](
https://github.com/jryans/rust/blob/cca3f8364911aa1400e2db1840104f5c6a51858d/src/librustdoc/html/render/mod.rs#L3518-L3532) of `render_assoc_items`.

Reply:
Got it, thanks - I missed that this is only skipping trait impls, not inherent impls. I'm not sure that behavior is correct, but it doesn't need to be fixed here.